pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,401 @@
+	c.podStore.Indexer = podInformer.GetIndexer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.updatePod,
+		DeleteFunc: c.deletePod,
+	})
+	buildInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addBuild,
+		UpdateFunc: c.updateBuild,
+	})
+
+	c.buildStoreSynced = buildInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *BuildPodController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the build store to sync before starting any work in this controller.
+	if !cache.WaitForCacheSync(stopCh, c.buildStoreSynced, c.podStoreSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down build pod controller")
+}
+
+// HandlePod updates the state of the build based on the pod state
+func (bc *BuildPodController) HandlePod(pod *kapi.Pod) error {
+	glog.V(5).Infof("Handling update of build pod %s/%s", pod.Namespace, pod.Name)
+	build, exists, err := bc.getBuildForPod(pod)
+	if err != nil {
+		glog.V(4).Infof("Error getting build for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		return err
+	}
+	if !exists || build == nil {
+		glog.V(5).Infof("No build found for pod %s/%s", pod.Namespace, pod.Name)
+		return nil
+	}
+
+	nextStatus := build.Status.Phase
+	currentReason := build.Status.Reason
+
+	if build.Status.Phase != buildapi.BuildPhaseFailed {
+		switch pod.Status.Phase {
+		case kapi.PodRunning:
+			// The pod's still running
+			build.Status.Reason = ""
+			build.Status.Message = ""
+			nextStatus = buildapi.BuildPhaseRunning
+
+		case kapi.PodPending:
+			build.Status.Reason = ""
+			build.Status.Message = ""
+			nextStatus = buildapi.BuildPhasePending
+			if secret := build.Spec.Output.PushSecret; secret != nil && currentReason != buildapi.StatusReasonMissingPushSecret {
+				if _, err := bc.secretClient.Secrets(build.Namespace).Get(secret.Name); err != nil && errors.IsNotFound(err) {
+					build.Status.Reason = buildapi.StatusReasonMissingPushSecret
+					build.Status.Message = buildapi.StatusMessageMissingPushSecret
+					glog.V(4).Infof("Setting reason for pending build to %q due to missing secret %s/%s", build.Status.Reason, build.Namespace, secret.Name)
+				}
+			}
+
+		case kapi.PodSucceeded:
+			build.Status.Reason = ""
+			build.Status.Message = ""
+			// Check the exit codes of all the containers in the pod
+			nextStatus = buildapi.BuildPhaseComplete
+			if len(pod.Status.ContainerStatuses) == 0 {
+				// no containers in the pod means something went badly wrong, so the build
+				// should be failed.
+				glog.V(2).Infof("Failing build %s/%s because the pod has no containers", build.Namespace, build.Name)
+				nextStatus = buildapi.BuildPhaseFailed
+			} else {
+				for _, info := range pod.Status.ContainerStatuses {
```

Comment:
This code was specifically added for a strange case where the pod was marked as succeeded but it had no containers.